### PR TITLE
Better errors descriptions for listen();

### PR DESCRIPTION
**Pull request by morozRed**
Summary:
- Add error-helper.js
- Add error-helper.test.js

### CODE REVIEW

**Comment by ai** [comment]
Good start. But if we want good docs for new users we should have better error message.

Let’s think together. Why user will get `Port already in use` error? 31337 and 1337 ports are very rare. So it is very rare case that some other non-Logux program use this port.

In my experience there are 2 most popular cases:

1. User have 2 Logux projects. He worked on first. And now manager asked him to fix some issue in second project. And developer forget to stop first Logux server before run second Logux server.
2. You deploy new Logux app to servers. But previous app was not killed correctly because of error in deploy script.

In all that cases “Try to use other port” will make only worse.

So, let’s change error text. You could try start with `port in use case`. Describe (with simple language) what is the problem. Describe possible reason why it happens. For each reason describe how to solve it.

You can start here. Just write it in comment.

**Comment by morozRed** [comment]
Thanks!
Ok, we can do something like this:
**EADDRINUSE**
```
ERROR   Port :3000 already in use.
HINT    Looks like Logux server already running on this port.
        Make sure there is no Logux servers running on this port.
```
**EACCES**
```
ERROR  You are not allowed to run server on this port.
HINT   Try to run server on port > 1024.
       You are currently using port: ${port}.
```

**Comment by ai** [comment]
Yeap, much better. But let’s write more details:

```
ERROR   Port :3000 already in use.
        Another Logux server or other app already running on this port.
        Maybe you didn’t not stop server from other project
        or previous version of this server was not killed.
```

Now you need to do same work with `EACCES`. Explain user why `<= 1024` posrt are different. What is the main reason of `EACCES` error?

**Comment by morozRed** [comment]
What about this?
**EACCES**
```
ERROR  You are not allowed to run server on this port.
HINT   You are not allowed to run your server on ports <= 1024
       Because they are privileged by security reasons.
       Try to run your server on ports > 1024.
```

**Comment by ai** [comment]
@morozRed do write docs without analysis ;). Tell me possible scenarios when users will have `EACCES` error.

**Comment by morozRed** [comment]
If we are talking only about listen() exceptions
1. User did something like like this in his app 
```
 app.listen({
    port: 1000
}) 
```
2. User is trying to access a file in a way forbidden by its file access permissions

**Comment by ai** [comment]
1. Why he or she will do it? I think most of users will copy code from docs ;).
2. Other reason — developer forget to write `sudo` to start server. I had this mistake for my own too :D.
3. `EACCES` could be for files too?!!! In this case we should detect port and file `EACCES`.

**Comment by morozRed** [comment]
Yes, you I just found it here: https://nodejs.org/api/errors.html#errors_error_code
I'll found a way we can detect EACCES type :) And then we will discuss error ` HINT ` again

**Comment by morozRed** [comment]
After little investigation I've realised that ` EACCES ` for files will not appear on listen() call. So we can finalize error message.
```
ERROR  You are not allowed to run server on this port
       Try to run server with sudo prefix or use port >= 1024
```

**Comment by ai** [comment]
Listen could load certificate files for TLS. Maybe we need good message for them.

**Comment by morozRed** [comment]
I've tested it and got this error:
``` 
ERROR  EACCES: permission denied, open '/server.key' at 2017-01-24 17:23:42

{ Error: EACCES: permission denied, open '/server.key'
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: '/server.key' }
```

**Comment by ai** [comment]
We need to cover this case too with good instructions.

**Comment by ai** [comment]
By the way, if `try to add sudo` is good recommendation. Running server as root is very very dangerous. Maybe we should ask “try to change user”?

**Comment by morozRed** [comment]
Ok, I'll think about this case with certificate.
Yes, probably it's better.

**Comment by ai** [comment]
OK, now let’s update text in PR and I will accept it.

**Comment by ai** [comment]
Few things to fix in code:

1. Use some real port instead of `undefined` in tests.
2. Move common code from `error-helper` and `reporter` to some common file.
3. You don’t need tricky `SEPARATOR` and `NEW_LINE` hack in `error-helper`.
4. Add real integration test. You can use some examples from `test/servers.js` tests.
5. Instead of `e.syscall === 'listen'` hack just use `try-catch`. Simple solutions is always better.

**Comment by morozRed** [comment]
Updates:
- updated errors messages
- add tests
- fix old tests
- add log-helper.js

File: server.js [hunk 1]
```diff
@@ -55,7 +56,11 @@ function Server (options) {
   var app = this
 
   function onError (e) {
-    app.reporter('runtimeError', app, undefined, e)
+    try {
+      process.stderr.write(errorHelper(e, app))
+    } catch (err) {
+      app.reporter('runtimeError', app, undefined, err)
+    }
```
Comment (ai) [inline server.js]: You misunderstand me. `onError` should contains only `app.reporter('runtimeError', app, undefined, e)`.

I told you to write try-catch in `listen()` method:

```
listen: function listen () {
  try {
    return BaseServer.prototype.listen.apply(this, arguments)
  } catch (e) {
    process.stderr.write(errorHelper(e, app))
  }
}
```

**Pull request merged**